Analysis of production chains (#186)
Leafwing-Studios/Emergence: design/src/production-chains/index.md

# Production chains

In a [factory builder](../high-level/genre-mechanics.md), resources are extracted, transformed and ultimately used.
They are extracted as [raw resources](raw-resources.md), are transformed into [intermediates](intermediates.md) are ultimately turned into useful [end products](end-products.md).

The paths that these **items** can travel on is dictated by the [recipes](recipes.md) available to the player.

## Analyzing production chains

When designing production chains for a factory builder (or frankly, any game with crafting),
the most essential thing to consider is the **recipe graph.**

You can construct the recipe graph by placing the items in the game as nodes of the graph, and then drawing an edge from each of the inputs in a recipe to its outputs.
We can also **color** this recipe graph into three parts that roughly correspond to raw resources, intermediates and end products.

> Raw resources have no parents, end products have no children*, and intemediates have both.

Reviewer: ```suggestion
Raw resources have no parents, end products have no children*, and intermediates have both.
```

As an aside, these are actually directed hypergraphs.
Hypergraphs may sound scary, but they're pretty simple: they're simply graphs that connect **nodes** to each other via **edges**, but the edges map betweeen groups of nodes rather than individual nodes.
"Directed" simply means that the edges have a direction to them: you can't turn electricity back into coal!

Unsurprisingly, this gets quite intimidating for most games!
Instead, it's more useful to:

1. Consider smaller parts of the graph at a time.
2. Reason abstractly about the properties of these graphs, and their effect on gameplay.

Let's go through some of the relevant properties of production graphs!

### Number of nodes (size)

This is a simple measure of the complexity of the game's crafting system.
If there are more items to find and make, there's more game to both learn and play.

> Increasing this value increases playtime and diversity at the cost of complexity and budget.

Reviewer: budget of? (I think you're aiming for complexity budget, but I suppose also financial/resource budget? more things to represent in the game visually/aurally? this could be alleviated to some extent by proc-gen I suppose)

Author: ```suggestion
Increasing this value increases playtime and diversity at the cost of complexity and development costs.
```

### Number of connected components

> Imagine the recipes as physically connecting various parts of your graph, and then try to pull it apart.

Reviewer: This is a little unclear. 

I think what you are trying to get at: "Imagine the recipes as inducing sub-graphs. How many of these sub-graphs are disconnected?"

possible normal-talk version: "Imagine the recipes as defining sub-graphs. How many of these sub-graphs share no vertices? 

But very likely I am misunderstanding.

Author: Yes, but the point is that this should be approachable. Number of connected components is a standard term.

How many parts can you split it into without breaking any of the edges?

The number of connected components is *very* high relative to the number of nodes in most farming games, survival crafters and other simple crafting systems.
This means that you don't need to reason too hard about what else you could use your resources for,
but limits the [cohesion](../glossary.md#cohesion) of the game's various systems and resource trees.

### Distance between nodes

This describes how many steps it takes to get from a desired set of raw resources to an end product.
When considered globally, this is the **graph diameter.**

This is often very low in simple crafting games, capping out at one or two.
Colony sims generally have chains of length one to three,
and even most factory builders end up with a chain length of one to four.

> This has a massive complexity cost, but adds a lot of [depth](../glossary.md#depth-and-complexity).

Reviewer: ```suggestion
Increasing this has a massive complexity cost, but adds a lot of [depth](../glossary.md#depth-and-complexity).
```

> Extending this further can be done, but requires some combination of:

Reviewer: ```suggestion
Increasing this well (from a game-design perspective) requires some combination of:
```

# Production chains

In a [factory builder](../high-level/genre-mechanics.md), resources are extracted, transformed and ultimately used.

> They are extracted as [raw resources](raw-resources.md), are transformed into [intermediates](intermediates.md) are ultimately turned into useful [end products](end-products.md).

Reviewer 2: ```suggestion
They are extracted as [raw resources](raw-resources.md), are transformed into [intermediates](intermediates.md) and ultimately turned into useful [end products](end-products.md).
```

The paths that these **items** can travel on is dictated by the [recipes](recipes.md) available to the player.

## Analyzing production chains

When designing production chains for a factory builder (or frankly, any game with crafting),
the most essential thing to consider is the **recipe graph.**

> You can construct the recipe graph by placing the items in the game as nodes of the graph, and then drawing an edge from each of the inputs in a recipe to its outputs.

Reviewer 2: How do you distinguish two recipes with multiple inputs in this graph? I.e., how can you tell which of the incoming edges "belong together"?
It would also be nice to include the _number_ of inputs and outputs that the recipes have.

I think one way to accomplish this would be to assign separate nodes to the recipes themselves.
In this case you can simply annotate the edges of the graph with the number of resources. The incoming edges of the recipes are the inputs & their numbers and the outgoing edge of the recipe goes to the item produced by the recipe. The number indicates how many items are produced.

We can also **color** this recipe graph into three parts that roughly correspond to raw resources, intermediates and end products.
Raw resources have no parents, end products have no children*, and intermediates have both.

As an aside, these are actually directed hypergraphs.

> Hypergraphs may sound scary, but they're pretty simple: they're simply graphs that connect **nodes** to each other via **edges**, but the edges map betweeen groups of nodes rather than individual nodes.

Reviewer 2: Why are the edges between groups of nodes in this case?

"Directed" simply means that the edges have a direction to them: you can't turn electricity back into coal!

Unsurprisingly, this gets quite intimidating for most games!
Instead, it's more useful to:

1. Consider smaller parts of the graph at a time.
2. Reason abstractly about the properties of these graphs, and their effect on gameplay.

Let's go through some of the relevant properties of production graphs!

### Number of nodes (vertex count)

This is a simple measure of the complexity of the game's crafting system.
If there are more items to find and make, there's more game to both learn and play.

Increasing this value increases playtime and diversity at the cost of complexity and development costs.

### Number of [connected components](https://en.wikipedia.org/wiki/Component_(graph_theory))

Imagine the recipes as physically connecting various parts of your graph, and then try to pull it apart.
How many parts can you split it into without breaking any of the edges?

> The number of connected components is *very* high relative to the number of nodes in most farming games, survival crafters and other simple crafting systems.

Reviewer 2: This kind of makes it sounds like the number of connected components is higher than the number of nodes, which is not possible.

Also, I think we should add how this relation is in most Factory builders. I assume you usually have a very small number of connected components?

Increasing this well (from a game-design perspective) requires some combination of:

- simpler, less effective early recipes to slowly teach the player
  - this is the strategy taken by the AngelBob's mod pack for Factorio: iron plates can be produced in many different ways, and only slowly becomes more complex
- good uses for intermediate steps in the chain, either as end products, or via heavy branching
  - this is also used in AngelBob's: sulfur is complex to make but exceedingly useful
- exceptionally good tooling that displays the entire production tree needed for each recipe
- exceptional patient and/or masochistic players
  - this is the strategy taken by [Pyanodon's mod pack for Factorio](https://www.reddit.com/r/factorio/comments/yd1i7l/finally_after_20_hours_of_work_i_present_the/)

### Number of incoming edges

How many different ways can I make this item?
In simple crafting games, this is almost always one, or a class of equivalent goods (different types of metal are common) can be substituted.

Even in factory builders, this can often be quite small: one is common for base gameplay, and even modded rarely has more than three or four options.

> When this is 1 everywhere, the recipe graph is a tree.

Reviewer 2: If we wanna be precise about this, it can also be a forest (or rather, a polyforest, because it's a directed graph), because the graph not necessarily connected: <https://en.wikipedia.org/wiki/Tree_(graph_theory)>

Constrictions can be quantified by examining the [vertex connectivity](https://en.wikipedia.org/wiki/K-vertex-connected_graph) and [edge connectivity](https://en.wikipedia.org/wiki/K-edge-connected_graph) of [neighborhoods](https://en.wikipedia.org/wiki/Neighbourhood_(graph_theory)).

## Production chains in *Emergence*

In *Emergence*, we're looking for production chains that are:

- interesting
- robust
- moderately complex

We have a lot of complexity elsewhere, as players must respond to changing conditions.
As a result, simply adding more and more items, and increasing the depth of the production chains is unlikely to be the right choice.

Instead, we want the following key properties:

- many cycles: everything should be able to be broken back down, either physically or via decomposition

> - heavy substitutability: powered by graph constrictions to key intermediates

Reviewer: Can this be explained further? I'm not sure I totally understand.
